File: src/invoker.rs
use super::command::*;

pub struct Invoker {
    execute_history: Vec<Box<dyn Command>>,
    undo_history: Vec<Box<dyn Command>>,
}

impl Invoker {
    pub fn new() -> Self {
        Invoker {
            execute_history: Vec::new(),
            undo_history: Vec::new(),
        }
    }

    pub fn execute_command(&mut self, mut command: Box<dyn Command>) {
        command.execute();
        //cannot mutate immutable variable `command`
        self.execute_history.push(command);
        self.undo_history.clear(); // 执行新命令后清空撤销历史
    }

    pub fn undo_command(&mut self) {
        if let Some(mut command) = self.execute_history.pop() {
            command.undo();
            self.undo_history.push(command);
        } else {
            println!("undo_command111111");
        }
    }

    pub fn redo_command(&mut self) {
        if let Some(mut command) = self.undo_history.pop() {
            command.redo();
            self.execute_history.push(command);
        } else {
            println!("redo_command2222");
        }
    }
}

File: invoker.cpp
#include "invoker.h"

Invoker::Invoker() {}

Invoker::~Invoker()
{
}

void Invoker::executeCommand(std::unique_ptr<Command> command)
{
    command->execute();
    undoStack_.push(std::move(command)); // 保存命令以便可以撤销
    while (!redoStack_.empty())
    {
        redoStack_.pop();
    }
}

void Invoker::undo()
{
    if (!undoStack_.empty())
    {
        std::unique_ptr<Command> command = std::move(undoStack_.top());
        undoStack_.pop();
        command->undo();
        redoStack_.push(std::move(command)); // 将命令移至重做栈
        std::cout << "撤销上一步操作" << std::endl;
    }
    else
    {
        std::cout << "undoStack 为空。 无法执行 undo 操作。" << std::endl;
    }
}

void Invoker::redo()
{
    if (!redoStack_.empty())
    {
        std::unique_ptr<Command> command = std::move(redoStack_.top());
        redoStack_.pop();
        command->redo();
        undoStack_.push(std::move(command)); // 重做后重新放回撤销栈
        std::cout << "重做上一步操作" << std::endl;
    }
    else
    {
        std::cout << "redoStack 为空。 无法执行 redo 操作。" << std::endl;
    }
}